### PR TITLE
Auto detect and set target feature CUDA on Linux

### PR DESCRIPTION
**Pull request by vmiheer**
This replaces default of OpenCl with cuda if cuda found.
I hope this is okay.

### CODE REVIEW

File: tutorial/lesson_12_using_the_gpu.cpp [hunk 1]
```diff
@@ -309,7 +309,20 @@ Target find_gpu_target() {
         target.set_feature(Target::Metal);
     }
 #else
-    if (dlopen("libOpenCL.so", RTLD_LAZY) != NULL) {
+    if (void *handle = dlopen("libcuda.so", RTLD_LAZY)) {
+        int numDevices = 0;
+        int (*cuDeviceGetCount)(int *);
```
Comment (steven-johnson) [inline tutorial/lesson_12_using_the_gpu.cpp]: Declare-and-initialize in one line is more appropriate.

**Comment by abadams** [comment]
This is a tutorial, which means we must minimize the number of lines of code that aren't directly teaching something, especially if they're hard to understand.

So I don't think this is a good change. If people want to use cuda they can uncomment the lines that enable cuda.

File: tutorial/lesson_12_using_the_gpu.cpp (hunk 1 above)
Comment (vmiheer) [reply tutorial/lesson_12_using_the_gpu.cpp]: Fixed. Thank you!

**Comment by vmiheer** [comment]
@abadams, I didn't know that. I though it would benefit people with cuda devices available.
The tutorial otherwise detects nvidia OpenCL but the nvprof doesn't even show the executed kernels (seems like nvidia dropped support for OpenCL in tools.) So I thought the change would be on the lines of how other api's are detected.
I leave the decision of whether to accept this change to you 😃.
I was also wondering existence of binary which sets HALIDE_TARGE/HALIDE_JIT_TARGET and runs the generator process as sub-process OR halide api to auto-populate target properties (refactoring of the code in the tutorial) if there is any value in it.

**Comment by steven-johnson** [comment]
I suspect we should probably close this, but will leave that up to @abadams

**Comment by abadams** [comment]
using host-cuda already sniffs the cuda host capability. I looked into trying to figure out an API to say "Does this machine support this target?", but I didn't get very far. It requires some design thought.

**Pull request closed without merging**